### ISSUE TITLE
qubes-builder make template-github curl broken - wrong environment variable handling

### ISSUE THREAD

**Issue by adrelanos** [QubesOS/qubes-issues#6737]
Related to https://github.com/QubesOS/qubes-issues/issues/3935.

> make DISTS_VM="whonix-gateway-15 whonix-workstation-15" template-github

Or.

> DEBUG=1 make DISTS_VM="whonix-gateway-15 whonix-workstation-15" template-github

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (56) Received HTTP code 403 from proxy after CONNECT
```

Nothing changed on my side. I can use github API manually.

If I manually set `data='-----BEGIN PGP SIGNED MESSAGE...'` and then manually run `curl --netrc-file /home/user/sourcesother/qubes-builder/.netrc_github https://api.github.com/repos/QubesOS/updates-status/issues/566/comments --data-binary @-`, it's functional. Therefore my Qubes API permission is functional too.

But when I run it with `make`, I get `curl: (56) Received HTTP code 403 from proxy after CONNECT`.

> env | grep -i http

```
REPO_PROXY=http://127.0.0.1:3142
COWBUILDER_PREFIX= env http_proxy=http://127.0.0.1:3142
```

Added `env` to qubes-builder `scripts/generate_build_github.sh`.

> cat build-logs/template-github-whonix-gateway-15.log | grep -i http

```
REPO_PROXY=http://127.0.0.1:3142
GIT_BASEURL=https://github.com
FETCH_CMD=http_proxy='http://127.0.0.1:3142' https_proxy='http://127.0.0.1:3142' curl --proto '=https' --proto-redir '=https' --tlsv1.2 --http1.1 -sSfL -o
https_proxy=http://127.0.0.1:3142
http_proxy=http://127.0.0.1:3142
ALL_PROXY=http://127.0.0.1:3142
COWBUILDER_PREFIX= env http_proxy=http://127.0.0.1:3142
```

Workaround for now:

> unset REPO_PROXY

Somehow qubes-builder seems to deduce `http_proxy` and so forth from `REPO_PROXY`? That seems wrong?

**Comment by DemiMarie** [comment]
> Somehow qubes-builder seems to deduce `http_proxy` and so forth from `REPO_PROXY`? That seems wrong?

That is intentional, so that downloading extra (not via git) sources via a proxy works.

> But when I run it with `make`, I get `curl: (56) Received HTTP code 403 from proxy after CONNECT`.

That seems like an issue with your proxy.

**Comment by adrelanos** [comment]
The name of the variable `REPO_PROXY ` indicates that it is just the APT (and arguably dnf) repository proxy. Otherwise why would there be a dedicated `REPO_PROXY` variable? That's a Qubes invention. Not used elsewhere.

If `http_proxy="$REPO_PROXY"` then there would be no need for a separate `REPO_PROXY` variable. Then one could just set `http_proxy`.

I don't think it's an issue with my proxy. It's apt-cacher-ng. It cannot handle TLS connections. [1] But that's alright since it's only used for non-TLS APT traffic (onion). The issue is imo that `http_proxy="$REPO_PROXY"` is wrong.

----

[1] [...] -> https://www.unix-ag.uni-kl.de/~bloch/acng/html/howtos.html#ssluse -> but besides the point.

**Comment by DemiMarie** [comment]
> The name of the variable `REPO_PROXY ` indicates that it is just the APT (and arguably dnf) repository proxy. Otherwise why would there be a dedicated `REPO_PROXY` variable? That's a Qubes invention. Not used elsewhere.

IIRC it is (or at least should be) a proxy for Git repositories, too.

> If `http_proxy="$REPO_PROXY"` then there would be no need for a separate `REPO_PROXY` variable. Then one could just set `http_proxy`.
> 
> I don't think it's an issue with my proxy. It's apt-cacher-ng. It cannot handle TLS connections. [1] But that's alright since it's only used for non-TLS APT traffic (onion). The issue is imo that `http_proxy="$REPO_PROXY"` is wrong.

I want to keep `REPO_PROXY` as the default for `http_proxy`, `https_proxy`, and friends, as having an easy way to set proxy settings is very convenient.  But it should not override values that have explicitly been set by the user.  

> [1] [...] -> https://www.unix-ag.uni-kl.de/~bloch/acng/html/howtos.html#ssluse -> but besides the point.

**Comment by marmarek** [comment]
I think @adrelanos has a point here. `REPO_PROXY` was introduced to use it for repository access, including a caching service, not a general purpose HTTP(S) proxy. Accepting https://github.com/QubesOS/qubes-builder/pull/165 was a mistake from my side.
In fact, the same applies to https://github.com/QubesOS/qubes-builder/pull/171 - it should be possible to set this separately (generic HTTP proxy for downloading various files, and specialized repository-access proxy).

**Comment by unman** [comment]
On Thu, Jun 24, 2021 at 05:14:53PM -0700, Marek Marczykowski-Górecki wrote:
> I think @adrelanos has a point here. `REPO_PROXY` was introduced to use it for repository access, including a caching service, not a general purpose HTTP(S) proxy. Accepting https://github.com/QubesOS/qubes-builder/pull/165 was a mistake from my side.
> In fact, the same applies to https://github.com/QubesOS/qubes-builder/pull/171 - it should be possible to set this separately (generic HTTP proxy for downloading various files, and specialized repository-access proxy).
> 

Thanks - I noticed this change in behaviour, which forces me to edit
builder.conf before make get-sources. 
It wasn't a welcome change: reverting both would be good for me. (I
have found a number of other changes ,in #154 for example, to have
unexpected effects.)

**Comment by DemiMarie** [comment]
@marmarek: Should `REPO_PROXY` be the default for `http_proxy` and `https_proxy`, or should they be entirely separate?

**Comment by marmarek** [comment]
> @marmarek: Should `REPO_PROXY` be the default for `http_proxy` and `https_proxy`, or should they be entirely separate?

No, they should be separate. Perhaps it should be the other way around: use `http_proxy`/`https_proxy` as default for `REPO_PROXY` (if it doesn't work this way already)?

**Comment by DemiMarie** [comment]
> > @marmarek: Should `REPO_PROXY` be the default for `http_proxy` and `https_proxy`, or should they be entirely separate?
> 
> No, they should be separate. Perhaps it should be the other way around: use `http_proxy`/`https_proxy` as default for `REPO_PROXY` (if it doesn't work this way already)?

Agreed.  That is consistent with standard behavior for \*nix command-line tools.

**Comment by unman** [comment]
On Fri, Jun 25, 2021 at 06:56:35AM -0700, Demi Marie Obenour wrote:
> > > @marmarek: Should `REPO_PROXY` be the default for `http_proxy` and `https_proxy`, or should they be entirely separate?
> > 
> > No, they should be separate. Perhaps it should be the other way around: use `http_proxy`/`https_proxy` as default for `REPO_PROXY` (if it doesn't work this way already)?
> 
> Agreed.  That is consistent with standard behavior for \*nix command-line tools.
> 

They should be entirely separate imo.

**Comment by adrelanos** [comment]
* Default `REPO_PROXY="$http_proxy"` - OKish, but wouldn't the application honor `http_proxy` anyhow? Therefore also pointless?
* Default `http_proxy="$REPO_PROXY"` - bad.

Setting `REPO_PROXY` makes sense if someone wants to use an apt cache (but then is responsible for additional proxyfication/socksification/torification/proxychain of that apt cache if that is desired).

Setting `http_proxy` makes sense if someone wants to use a http (only? Not socks?) proxy.

Just someone uses both `http_proxy` and `REPO_PROXY`, then one shouldn't assume that `REPO_PROXY` uses `http_proxy` (a proxy chain) since I guess that would be hard to implement at Qubes builder level.

**Comment by unman** [comment]
On Fri, Jun 25, 2021 at 08:45:48AM -0700, Demi Marie Obenour wrote:
> What about ???If it is set, `REPO_PROXY` overrides the standard proxy environment variables, but _only_ for requests made by package managers.???
> 

I dont see an advantage in this: They are separate.
Set a separate environment proxy if you will, and leave REPO_PROXY as it was.

**Comment by DemiMarie** [comment]
This is still a bug, but I am not sure if it is worth fixing in the current builder, since the new builder (currently being written by @fepitre) will replace everything.

**Comment by ben-grande** [comment]
Builder v1 fallen into disuse, last update was 2 years ago. Builder v2 has it fixed, only pacman uses [REPO_PROXY](https://github.com/search?q=repo%3AQubesOS%2Fqubes-builderv2%20REPO_PROxY&type=code), it uses `http_proxy='${REPO_PROXY}'`
